Review agents upgrade API integration tests

Right now, our `test_agent_PUT_endpoints.tavern.yaml` API integration test is randomly failing.

The following is a test output where this random error appears:

<details>

<summary>output</summary>

```
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.8.5', 'Platform': 'Linux-5.4.0-1029-aws-x86_64-with-glibc2.29', 'Packages': {'pytest': '5.4.3', 'py': '1.9.0', 'pluggy': '0.13.1'}, 'Plugins': {'tavern': '1.0.0', 'metadata': '1.10.0', 'html': '2.0.1'}}
rootdir: /tmp/Test_integration_endpoints_B1543_test_files/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, metadata-1.10.0, html-2.0.1
collecting ... collected 10 items

test_agent_PUT_endpoints.tavern.yaml::PUT /agents/group PASSED           [ 10%]
test_agent_PUT_endpoints.tavern.yaml::PUT /groups/{group_id}/configuration PASSED [ 20%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/group/{group_id}/restart PASSED [ 30%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/reconnect PASSED       [ 40%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/restart PASSED         [ 50%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/{agent_id}/group/{group_id} PASSED [ 60%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/{agent_id}/restart PASSED [ 70%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/node/{node_id}/restart XFAIL [ 80%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/upgrade FAILED         [ 90%]
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/{agent_id}/upgrade_custom XFAIL [100%]

=================================== FAILURES ===================================
_ /tmp/Test_integration_endpoints_B1543_test_files/api/test/integration/test_agent_PUT_endpoints.tavern.yaml::PUT /agents/upgrade _
Format variables:
  protocol:s = 'https'
  host:s = 'localhost'
  port:d = '55000'
  test_login_token = 'eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjM2MTEzMjE5LCJleHAiOjE2MzYxMTQxMTksInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbMV0sInJiYWNfbW9kZSI6IndoaXRlIn0.AC9FcHIHKDPoCW3XquMLCng1k4REu3xg2C7OlMYkTQlheSArpns97BhFoEp4QpCNHSd8Osiw9mMEb8tHidZCF8AkAQrFgcvkssY00MY3vg_OLNn5F4Df9iIPfme0UCXwvJseRSf2hguA0hPR-MBaNPmJd0qp-fiID6TxZ_LofLGwgq66'

Source test stage (line 933):
  - name: Try to upgrade an agent that is updated to the latest version
    request:
      verify: False
      url: "{protocol:s}://{host:s}:{port:d}/agents/upgrade"
      method: PUT
      headers:
        Authorization: "Bearer {test_login_token}"
      params:
        agents_list: '001'
    response:
      status_code: 200
      json:
        data:
          affected_items: [ ]
          total_affected_items: 0
          total_failed_items: 1
          failed_items:
            - error:
                code: 1822
                message: !anystr
                remediation: !anything
              id:
                - '001'
        message: !anystr


Formatted stage:
  name: Try to upgrade an agent that is updated to the latest version
  request:
    headers:
      Authorization: Bearer eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjM2MTEzMjE5LCJleHAiOjE2MzYxMTQxMTksInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbMV0sInJiYWNfbW9kZSI6IndoaXRlIn0.AC9FcHIHKDPoCW3XquMLCng1k4REu3xg2C7OlMYkTQlheSArpns97BhFoEp4QpCNHSd8Osiw9mMEb8tHidZCF8AkAQrFgcvkssY00MY3vg_OLNn5F4Df9iIPfme0UCXwvJseRSf2hguA0hPR-MBaNPmJd0qp-fiID6TxZ_LofLGwgq66
    method: PUT
    params:
      agents_list: '001'
    url: 'https://localhost:55000/agents/upgrade'
    verify: false
  response:
    json:
      data:
        affected_items: []
        failed_items:
        - error:
            code: 1822
            message: !anystr ''
            remediation: !anything ''
          id:
          - '001'
        total_affected_items: 0
        total_failed_items: 1
      message: !anystr ''
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Try to upgrade an agent that is updated to the latest version' failed:
    - List item(s) not present in response: [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}]
------------------------------ Captured log call -------------------------------
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '400'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
ERROR    tavern.response.base:base.py:41 List item(s) not present in response: [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}]
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'data': {'af... was created'} == {'data': {'af...7f94a3a91580>}
  Differing items:
  {'message': 'No upgrade task was created'} != {'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a91580>}
  {'data': {'affected_items': [], 'failed_items': [{'error': {'code': 1707, 'message': 'Cannot send request, agent is no...formation on registering and connecting agents'}, 'id': ['001']}], 'total_affected_items': 0, 'total_failed_items': 1}} != {'data': {'affected_items': [], 'failed_items': [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel o...ader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}], 'total_affec...
  
  ...Full output truncated (31 lines hidden), use '-vv' to show

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'affected_it...led_items': 1} == {'affected_it...led_items': 1}
  Omitting 3 identical items, use -vv to show
  Differing items:
  {'failed_items': [{'error': {'code': 1707, 'message': 'Cannot send request, agent is not active', 'remediation': 'Plea...manual/agents/agent-connection.html to obtain more information on registering and connecting agents'}, 'id': ['001']}]} != {'failed_items': [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}]}
  Full diff:
    {
     'affected_items': [],
  -  'failed_items': [{'error...
  
  ...Full output truncated (20 lines hidden), use '-vv' to show

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert [{'error': {'...id': ['001']}] == [{'error': {'...id': ['001']}]
  At index 0 diff: {'error': {'code': 1707, 'message': 'Cannot send request, agent is not active', 'remediation': 'Please, check non-active agents connection and try again. Visit https://documentation.wazuh.com/4.3/user-manual/registering/index.html and https://documentation.wazuh.com/4.3/user-manual/agents/agent-connection.html to obtain more information on registering and connecting agents'}, 'id': ['001']} != {'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f9...
  
  ...Full output truncated (19 lines hidden), use '-vv' to show

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/response/base.py", line 88, in recurse_check_key_match
    check_keys_match_recursive(expected_block, block, [], strict)
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 462, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 462, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 512, in check_keys_match_recursive
    raise exceptions.KeyMismatchError(msg) from e
tavern.util.exceptions.KeyMismatchError: List item(s) not present in response: [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}]
=============================== warnings summary ===============================
/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/hooks.py:83
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/hooks.py:83: PytestDeprecationWarning: direct construction of YamlFile has been deprecated, please use YamlFile.from_parent
    return YamlFile(path, parent)

/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:233: 10 tests with warnings
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:233: PytestDeprecationWarning: direct construction of YamlItem has been deprecated, please use YamlItem.from_parent
    item = YamlItem(test_spec["test_name"], self, test_spec, self.fspath)

/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:61
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:61: PytestUnknownMarkWarning: Unknown pytest.mark.base_tests - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    pytest_marks.append(getattr(pytest.mark, m))

-- Docs: https://docs.pytest.org/en/latest/warnings.html
- generated html file: file:///tmp/Test_integration_endpoints_B1543_test_files/api/test/integration/Test_integration_endpoints_B1543_test_agent_PUT_endpoints.html -
=========================== short test summary info ============================
FAILED test_agent_PUT_endpoints.tavern.yaml::PUT /agents/upgrade
======= 1 failed, 7 passed, 2 xfailed, 12 warnings in 786.37s (0:13:06) ========
```

</details>

As it can be seen, the error is in the test case `PUT /agents/upgrade`:

```
test_agent_PUT_endpoints.tavern.yaml::PUT /agents/upgrade FAILED         [ 90%]
```

When taking a look to the traceback, we see the following:

```
  {'failed_items': [{'error': {'code': 1707, 'message': 'Cannot send request, agent is not active', 'remediation': 'Plea...manual/agents/agent-connection.html to obtain more information on registering and connecting agents'}, 'id': ['001']}]} != {'failed_items': [{'error': {'code': 1822, 'message': <tavern.util.loader.StrSentinel object at 0x7f94a3a916a0>, 'remediation': <tavern.util.loader.AnythingSentinel object at 0x7f94a46e0430>}, 'id': ['001']}]}
```

Agent 001 is not active when we try to upgrade it. We don't expect the error code 1707 which is the one specifying the agent was not active, and that is why the test fails.

In this issue, review and fix the API integration test. Also note that when this error appears, the test cases: `PUT /agents/node/{node_id}/restart` and `PUT /agents/{agent_id}/upgrade_custom`, which are marked with the `xfail` mark, are also failing. This could mean that agent 001 has status `pending` by the time we try to upgrade it. Therefore, the error could be the same as the one reported [here](https://github.com/wazuh/wazuh/issues/10162).


## Checks

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)
<!-- If changes are made to any of the following components, uncomment the corresponding line 
- [ ] **Integration tests** without failures for API configuration (`/wazuh-qa/tests/integration/test_api/test_config/`)
- [ ] **System tests** for agent enrollment process (`/wazuh-qa/tests/system/test_cluster/test_agent_enrollment`)
- [ ] **System tests** for agent info sync process in cluster (`/wazuh-qa/tests/system/test_cluster/test_agent_info_sync`)
- [ ] **System tests** for agent key polling (`/wazuh-qa/tests/system/test_cluster/test_agent_key_polling`)
- [ ] **System tests** for JWT invalidation (`/wazuh-qa/tests/system/test_jwt_invalidation`)
-->

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

--- a/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
+++ b/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
@@ -846,7 +846,7 @@ stages:
 test_name: PUT /agents/upgrade
 
 marks:
-  - xfail # Review agents upgrade API integration tests - https://github.com/wazuh/wazuh/issues/10791
+  - xfail # Agents stay with "pending" status for too long - https://github.com/wazuh/wazuh/issues/10162
 
 stages:
 

